pkg/storage/bloom/v1/bloom_tester.go

```diff
@@ -35,6 +35,39 @@ func (b BloomTests) MatchesWithPrefixBuf(bloom filter.Checker, buf []byte, prefi
 	return true
 }
 
+// ExtractTestableLineFilters extracts all line filters from an expression
+// that can be tested against a bloom filter. This will skip any line filters
+// after a line format expression. A line format expression might add content
+// that the query later matches against, which can't be tested with a bloom filter.
+// E.g. For {app="fake"} |= "foo" | line_format "thisNewTextShouldMatch" |= "thisNewTextShouldMatch"
+// this function will return only the line filter for "foo" since the line filter for "thisNewTextShouldMatch"
+// wouldn't match against the bloom filter but should match against the query.
+func ExtractTestableLineFilters(expr syntax.Expr) []syntax.LineFilterExpr {
+	if expr == nil {
+		return nil
+	}
+
+	var filters []syntax.LineFilterExpr
+	var lineFmtFound bool
+	visitor := &syntax.DepthFirstTraversal{
+		VisitLineFilterFn: func(v syntax.RootVisitor, e *syntax.LineFilterExpr) {
+			if e != nil && !lineFmtFound {
+				filters = append(filters, *e)
+			}
+		},
+		VisitLineFmtFn: func(v syntax.RootVisitor, e *syntax.LineFmtExpr) {
+			if e != nil {
```

Comment:
Is it possible that the visitor visits a `LineFmtExpr` that is `nil`?

Reply:
Probably not but not sure tbh. Still I think it's a good practice to check ptrs. Happy to remove it tho.